### ISSUE TITLE
ANSI escape characters in kubectl output are not being filtered

### ISSUE THREAD

**Issue by g3rzi** [kubernetes/kubernetes#101695]
It is a security issue, but after contacting security@kubernetes.io, Tim and the team confirmed that they are comfortable posting it publicly.

#### What happened: 
Kubernetes doesn't sanitize the 'message' field in the Event JSON objects.  
Notice that this is relevant only to JSON objects, not YAML objects. 
By creating new event, we can insert ANSI escape characters inside the "message" field, like:
```
"\u001b[2J\u001b[3J\u001b[1;1H\u001b[m spoofed \u001b[60C\u001b[1;0m. Done"
```

This an example of such JSON request:
```
{
    "apiVersion": "v1",
    "involvedObject": {
        "kind": "Node",
        "name": "node01",
        "uid": "node01"
    },
    "kind": "Event",
    "message": "\u001b[2J\u001b[3J\u001b[1;1H\u001b[m spoofed \u001b[60C\u001b[1;0m. Done",
    "metadata": {
        "name": "spoofEvent",
        "namespace": "default"
    },
    "source": {
        "component": "kubelet",
        "host": "node01"
    },
    "type": "Normal"
}
```

The codes:
`\u001b[2J`   -> Clean the screen and history
`\u001b[3J`   -> Clean the entire screen and delete all lines saved in the scrollback buffer 
`\u001b[1;1H` -> Moves the cursor position to row 1, column 1 (beginning).
`\u001b[m`    -> Set the colors
`\u001b[60C`  -> Move the cursor forward 60 steps
`\u001b[1;1m` -> Set the text colors to white

The result is that the text was spoofed, and we could spoof the events, create hidden events, or hide other events.

#### What you expected to happen: 
The ANSI escape characters will be filtered so they couldn't affect the terminal (i.e. using embeded ANSI colors won't do anything to the terminal).
Or maybe some message that says that you can't use ANSI escape characters.

#### How to reproduce it (as minimally and precisely as possible):

1. Run this code:
```

kubectl create -f - <<EOF
{
    "apiVersion": "v1",
    "involvedObject": {
        "kind": "Node",
        "name": "node01",
        "uid": "node01"
    },
    "kind": "Event",
    "message": "\u001b[2J\u001b[3J\u001b[1;1H\u001b[m spoofed \u001b[60C\u001b[1;0m. Done",
    "metadata": {
        "name": "spoofEvent",
        "namespace": "default"
    },
    "source": {
        "component": "kubelet",
        "host": "node01"
    },
    "type": "Normal"
}
EOF
```
It will create a new event.

2. Run `kubectl get events`, you will see that the screen was clear, you will get a "spoof" message, and all the rest events or columns were gone.

#### Anything else we need to know?:

It might look like a low severity issue, but there are other variety of things we can do, from DoS by using colors to hide all the events, changing the title of the terminal window, and spoof the data.

It can affect other systems that are using Kubernetes events, such as monitoring applications. It doesn't have to be only the Kubernetes events. There might be other vulnerable objects that we didn't find or other systems that create new objects that count on this mechanism.

ANSI escape characters were used to abuse terminals emulators and even cause code execution if the terminal is vulnerable (like CVE-2003-0069).

#### Environment:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.4", GitCommit:"e87da0bd6e03ec3fea7933c4b5263d151aafd07c", GitTreeState:"clean", BuildDate:"2021-02-18T16:12:00Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-13T13:20:00Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: 
- OS (e.g: `cat /etc/os-release`):
```
NAME="Ubuntu"
VERSION="16.04.7 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.7 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```
- Kernel (e.g. `uname -a`):
```
Linux manager1 4.15.0-140-generic #144~16.04.1-Ubuntu SMP Fri Mar 19 21:24:12 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

- Install tools: minikube
```
minikube version: v1.19.0
commit: 15cede53bdc5fe242228853e737333b09d4336b5
```

- Network plugin and version (if this is a network-related bug): 
- Others: we also reproduced it in Kubernetes (not minikube) version 1.18

**Comment by pacoxu** [comment]
I built an image `daocloud.io/daocloud/test-escape:0.1`.
- The log contains the escape chars as well.

`kubectl logs` would have the same problem as well.
/sig cli

**Comment by g3rzi** [comment]
Any update on this issue?

**Comment by g3rzi** [comment]
Any updates?

**Comment by soltysh** [comment]
@g3rzi apologies, this slipped my attention, I'll have a look at it later this week 
/assign

**Comment by g3rzi** [comment]
Hi @soltysh, did you have time to look at this issue?

**Comment by g3rzi** [comment]
Any update?

**Comment by g3rzi** [comment]
Hi @soltysh, did you have time to look at this issue?

**Comment by palnabarun** [comment]
cc: @RinkiyaKeDad

**Comment by RinkiyaKeDad** [comment]
I can also help with this 🙂

/assign

**Comment by soltysh** [comment]
I spoke with @RinkiyaKeDad, he'll be looking at the this. In short I'm proposing to update https://github.com/kubernetes/kubernetes/blob/758ad0790ceae9e7be0554637c0fd623ff02e45f/staging/src/k8s.io/cli-runtime/pkg/printers/json.go#L37-L82 to properly handle escape chars handling. Looking at [JSON spec](https://datatracker.ietf.org/doc/html/rfc8259#section-7) I'd suggest using `strconv.QuoteToASCII` which was also proposed in [this golang issue](https://github.com/golang/go/issues/39137).

**Comment by tallclair** [comment]
IMO this should apply to all arbitrary strings output to a terminal by command line tools. It's probably simpler to just put a catch-all around kubectl output to escape control characters, but I'm not sure if kubectl intentionally uses control characters anywhere. I am neutral on whether the escaping should still be applied when the kubectl output is piped to another command.

I do NOT think the escaping should be applied server side. This is only an issue for tools outputting the data to a terminal.

**Comment by BenTheElder** [comment]
> but I'm not sure if kubectl intentionally uses control characters anywhere.

kubectl uses ansi escape codes (color, bold ...) when printing warnings since https://github.com/kubernetes/kubernetes/pull/73032

**Comment by RSAlderman** [comment]
@RinkiyaKeDad that related issue https://github.com/kubernetes/kubernetes/pull/104962#issuecomment-1005426583 closed waiting on solution design

Am I correct that there is currently no fix for this issue, assigned `CVE-2021-25743`?

**Comment by RinkiyaKeDad** [comment]
@RSAlderman from what I'm aware of, https://github.com/kubernetes/kubernetes/pull/104962 was the only PR out to fix this.

**Comment by RSAlderman** [comment]
@RinkiyaKeDad thanks for the prompt response. You said last week in #104962
> The issue still needs to be fixed ...
> will open a new PR once we've decided how to fix it exactly.

I presume that PR will be referenced from this issue when it is produced and then backported to 1.22 and 1.21 too?

**Comment by RinkiyaKeDad** [comment]
> and then backported to 1.22 and 1.21 too

Sorry I'm afraid I don't know if this will happen or not. Let's wait for what others have to say 🙂
Also, just as an fyi, I'm not working on this currently and I don't think anyone else is too.

**Comment by liggitt** [comment]
As noted in https://github.com/kubernetes/kubernetes/issues/107617, this also applies to display of annotations

**Comment by p-rog** [comment]
Are there any news regarding potential patch to this vulnerability?
It seems that some fix in kubectl is the best option right now.

**Comment by sayananavya** [comment]
Team, Does anyone has fix for this vulnerability? What packages need to be changed/updated??

Thanks in advance.

**Comment by pacoxu** [comment]
@dgl is working on this. #112553 is open.

**Comment by DamianSawicki** [comment]
@buckaroogeek [asked](https://github.com/kubernetes/kubernetes/pull/112553#issuecomment-1329408815) about it already in the PR, but got no response for 11 months, so I'll ask again here: will the fix be applied to 1.25 and 1.24? @dgl @g3rzi

**Comment by buckaroogeek** [comment]
@DamianSawicki - thanks for checking again. I suspect this is almost moot. 1.24 is no longer supported and 1.25 end-of-life was supposed to be 2023.10.27.

**Comment by dgl** [comment]
It's unlikely this will be backported now. Note that a 1.25 cluster can be used with kubectl 1.26 per [version skew policy](https://kubernetes.io/releases/version-skew-policy/#kubelet-1) and 1.24 on the cluster is likely fine too in this case, so given those are out of support it doesn't make sense to backport, just use a newer kubectl.

**Comment by DamianSawicki** [comment]
I see, thank you both for the answers!
